### PR TITLE
Respect privacy and limitations of use and import statements when resolving methods

### PR DESCRIPTION
**Pull request by lydia-duncan**
Removes the support for ignoring the privacy and limitations of use and import statements
when resolving methods.  While the implementation is slightly more complex as a result, it
simplifies the explanation of when symbols can and cannot be found.

Resolves Cray/chapel-private#1423

<Details>

Updates a comment to reflect the new behavior.

Makes getVisibleMethodsImpl pay attention to whether a scope has been entered through
a use/import chain or by traversing upwards from the call site.  This allows us to distinguish
between private uses that we should follow (because we are in the scope where they occur)
and private uses that we should ignore.  Pulls the act of checking the symbols defined
directly in a scope into a helper function, as well as traversing the use list.

Adjusts lookAtTypeFirst to also jump to the point where a parent type was defined.  The
traversal of private uses had allowed us to find inherited methods, but we need to explicitly
jump to those scopes now that such uses are respected.  Also track whether the scope we
jump to is the same as the scope we are currently in - if not, we should respect the privacy
of any use and import statements we find while there.

Remove isScopeVisibleForMethods and its helper map.  We can now rely on normal visibility
computation.

A couple of modules needed updates for this new behavior - BlockDist and StencilDist were
relying on the cast function for c_void_ptr just being visible and now explicitly use CPtr
(where it is defined), and the TOML module now has a use of IO in order to access the format
method on strings.

Testing
-------

Adds a new test to lock in that we find all inherited methods, even if the inheritance hierarchy
is deeper than just two levels.  Updates the .bad output for one test due to the altered order in
which we find its candidates.  Renames and updates the behavior for
qualified-diffModule-okay*.chpl, because the behavior it tests is now a failure mode.  While
there, fixes a related test to better reflect its naming (and cover the case it was supposed to
cover, which was using an `except *`).  Converts canSeeSecondaryMethod2.chpl into a future
until we add support for listing a type in an only or except list and having it impact the visibility
of methods on that type visible to that scope.  And finally updates the behavior of a test
tracking what happens when a secondary method is behind a private use.
</details>

Passed a full paratest with futures

### CODE REVIEW

**Comment by lydia-duncan** [comment]
@vasslitvinov - would you review this, especially since it probably will have conflicts with your work?

File: compiler/resolution/visibleFunctions.cpp [hunk 1]
```diff
@@ -415,26 +411,50 @@ static void getVisibleMethodsVI(const char* name, CallExpr* call,
                                 Vec<FnSymbol*>&       visibleFns)
 {
   INT_ASSERT(visited->empty() == !visInfo->inPOI()); //fyi
+  BlockStmt* block = getVisibilityScope(call);
```
Comment (mppf) [inline compiler/resolution/visibleFunctions.cpp:414]: Move this variable inside the conditional below since it is only used there?

File: compiler/resolution/visibleFunctions.cpp [hunk 2]
```diff
@@ -449,131 +469,161 @@ static void lookAtTypeFirst(const char* name, CallExpr* call,
   }
   Type* t = typeActual->getValType();
 
+  BlockStmt* typeScope = getVisibilityScope(t->symbol->defPoint);
+  // If we are already in the scope where the type is defined, we need to
+  // indicate that.  However, if we are in a different scope (even if we're in a
+  // subscope of where the type is defined), then we should treat this check as
+  // though we are following an explicit use statement, and not follow private
+  // use and import statements.
```
Comment (mppf) [inline compiler/resolution/visibleFunctions.cpp:477]: This comment is just a copy of the one above, maybe write something shorter this time?
Either way I find this comment more confusing than the code, personally.

File: compiler/resolution/visibleFunctions.cpp [hunk 3]
```diff
@@ -449,131 +469,161 @@ static void lookAtTypeFirst(const char* name, CallExpr* call,
...
-    if (call->id == breakOnResolveID) {
-      getVisibleFnsShowBlock("methods", block, instantiationPt);
-    }
+  const bool firstVisit = (visited.find(block) == visited.end());
+  if (!firstVisit && inUseChain) {
+    // We've seen this block already, but we just found it again.  If we found
+    // it again from another use chain, just return.  If we found it from going
+    // up in scope from the call site, that means that we may have skipped
+    // private uses, so we should go through only the private uses - not in a
+    // use chain.
```
Comment (mppf) [inline compiler/resolution/visibleFunctions.cpp:575]: I found the comment hard to follow. Could it be:
```
    // We've seen this block already, but we just found it again from a use chain.
    //   * If we found it earlier from another use chain, no further action is needed, so return
   //    * If we found it earlier from going up in scope from the call site, no further action is needed, so return
```

(You could say more about private uses in there but I found it confusing - what is a private use not in a use chain?).

File: compiler/resolution/visibleFunctions.cpp [hunk 4]
```diff
@@ -449,131 +469,161 @@ static void lookAtTypeFirst(const char* name, CallExpr* call,
...
-          }
+  if (firstVisit && block->modRefs != NULL) {
+    for_actuals(expr, block->modRefs) {
+      SymExpr* se = toSymExpr(expr);
+      INT_ASSERT(se);
+      if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
+        if (mod->isVisible(call)) {
+          // Treat following a module reference as though it was a use statement
+          // for the purpose of determining whether we can follow private uses
+          // and imports
```
Comment (mppf) [inline compiler/resolution/visibleFunctions.cpp:603]: Can you show a code sample of what this is trying to support? Is it things like `OtherModule.foo()` ?

**Review by mppf — APPROVED** [review]
Generall, looks good to me!

@vasslitvinov  should also look this over when convenient

File: compiler/resolution/visibleFunctions.cpp (hunk 3 above)
Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:575]: The boolean doesn't track how we found the scope previously, it tracks how we are currently finding the scope.  There's three scenarios:
- found the scope the first time by traversing a use
  - found the scope again by also traversing a use
  - found the scope again by going up in scope from the call site (this is the only case where we care that we found the scope again)
- found the scope the first time by going up in scope from the call site, and again by traversing a use
(it is not possible to find the scope twice when both times were by going up scope from the call site - is what I was going to type, but the lookAtTypeFirst stuff changes that, now that I think of it . . .)

We don't traverse private use statements unless we found the scope with the use statements by going up in scope from the call site.  And we don't traverse scopes we've already seen before unless there's a reason for us to believe we'll find different symbols this time around.

File: compiler/resolution/visibleFunctions.cpp (hunk 4 above)
Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:603]: Yeah, that's basically it

File: compiler/resolution/visibleFunctions.cpp (hunk 3 above)
Comment (mppf) [reply compiler/resolution/visibleFunctions.cpp:575]: Seems like most of what you just posted would make a great and more understandable comment there :)

File: compiler/resolution/visibleFunctions.cpp (hunk 2 above)
Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp]: How does it look now?

Comment (mppf) [reply compiler/resolution/visibleFunctions.cpp]: sounds good
